### PR TITLE
Add useful Debug impl to PropertyBag and ConfigBag

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- it's hard to debug what properties are set, especially for layered configuration

## Description
Add Debug Info for PropertyBag and Config bag:
```
PropertyBag { contents: ["aws_smithy_http::property_bag::test::test_extensions::MyType"] }
```

```

ConfigBag {
    layers: [
        Layer {
            name: "c",
            properties: [
                "aws_smithy_runtime_api::config_bag::Value<aws_smithy_runtime_api::config_bag::test::layered_property_bag::Prop3>",
                "aws_smithy_runtime_api::config_bag::Value<aws_smithy_runtime_api::config_bag::test::layered_property_bag::Prop4>",
            ],
        },
        Layer {
            name: "b",
            properties: [
                "aws_smithy_runtime_api::config_bag::Value<aws_smithy_runtime_api::config_bag::test::layered_property_bag::Prop3>",
                "aws_smithy_runtime_api::config_bag::Value<aws_smithy_runtime_api::config_bag::test::layered_property_bag::Prop2>",
            ],
        },
        Layer {
            name: "a",
            properties: [
                "aws_smithy_runtime_api::config_bag::Value<aws_smithy_runtime_api::config_bag::test::layered_property_bag::Prop1>",
            ],
        },
        Layer {
            name: "base",
            properties: [],
        },
    ],
}
```

There is still some work to do, but this is a start

## Testing
- [x] UT

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-http/src/property_bag.rs [hunk 1]
```diff
@@ -225,22 +243,30 @@ impl From<PropertyBag> for SharedPropertyBag {
...
+        assert_eq!(property_bag.remove::<i32>(), Some(5i32));
+        assert!(property_bag.get::<i32>().is_none());
 
-    assert_eq!(extensions.get::<bool>(), None);
-    assert_eq!(extensions.get(), Some(&MyType(10)));
+        assert_eq!(property_bag.get::<bool>(), None);
+        assert_eq!(property_bag.get(), Some(&MyType(10)));
+        assert_eq!(
+            format!("{:?}", property_bag),
+            r#"PropertyBag { contents: ["aws_smithy_http::property_bag::test::test_extensions::MyType"] }"#
```
Comment (jjant) [inline rust-runtime/aws-smithy-http/src/property_bag.rs:269]: `std::any::type_name`'s output is essentially unstable, so I'm not sure if we want this test, see https://doc.rust-lang.org/std/any/fn.type_name.html#note.

File: rust-runtime/aws-smithy-runtime-api/src/config_bag.rs [hunk 2]
```diff
@@ -258,6 +292,7 @@ mod test {
         assert!(final_bag.get::<Prop2>().is_some());
         // we unset prop3
         assert!(final_bag.get::<Prop3>().is_none());
+        println!("{:#?}", final_bag);
```
Comment (jjant) [inline rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:295]: Did you want this here?

File: rust-runtime/aws-smithy-http/src/property_bag.rs [hunk 3]
```diff
@@ -106,7 +110,12 @@ impl PropertyBag {
     pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
         self.map
             .get(&TypeId::of::<T>())
-            .and_then(|boxed| (&**boxed as &(dyn Any + 'static)).downcast_ref())
+            .and_then(|(_, boxed)| (&**boxed as &(dyn Any + 'static)).downcast_ref())
+    }
+
+    /// Returns an iterator of the types contained in this PropertyBag
+    pub fn contents(&self) -> impl Iterator<Item = &'static str> + '_ {
```
Comment (jjant) [inline rust-runtime/aws-smithy-http/src/property_bag.rs:117]: Do we want this to be `pub`?

Comment (rcoh) [reply rust-runtime/aws-smithy-http/src/property_bag.rs:117]: yeah we use it in config_bag, added a note about stability

File: rust-runtime/aws-smithy-runtime-api/src/config_bag.rs (hunk 2 above)
Comment (rcoh) [reply rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:295]: yeah, just verifying how the bag looks for debug purposes

**Review by Velfi — APPROVED** [review]

**Review by jdisanti — APPROVED** [review]
Nice! This is awesome

**Review by ysaito1001 — APPROVED** [review]

**Pull request merged**